### PR TITLE
Cut short recursion that can occur (by spec!) in the better betterness code.

### PR DESCRIPTION
**Pull request by gafter**
This handles the simplest cases that are likely to arise in practice. We simply limit
direct recursion in "better conversion target" to a depth of 100.
See #12061 for the related spec bug. The spec still must be adjusted to say what should happen here,
and the implementation will need to be changed to follow that new specification.

@khyperia @dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 1]
```diff
@@ -2015,13 +2015,30 @@ public override BoundNode VisitReturnStatement(BoundReturnStatement node)
             }
         }
 
-        private BetterResult BetterConversionTarget(TypeSymbol type1, TypeSymbol type2, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        private BetterResult BetterConversionTarget(
+            TypeSymbol type1,
+            TypeSymbol type2,
+            ref HashSet<DiagnosticInfo> useSiteDiagnostics,
+            int betterConversionTargetRecursionLimit = 100)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: We should make this a constant and a non-default argument.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: Entries to these functions from outside the recursion (i.e. other places in the compiler) should not have to be aware of this parameter. That is why I prefer this pattern.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 2]
```diff
@@ -2015,13 +2015,30 @@ public override BoundNode VisitReturnStatement(BoundReturnStatement node)
...
-        private BetterResult BetterConversionTarget(BoundExpression node, TypeSymbol type1, Conversion conv1, TypeSymbol type2, Conversion conv2, ref HashSet<DiagnosticInfo> useSiteDiagnostics, out bool okToDowngradeToNeither)
+        private BetterResult BetterConversionTarget(
+            BoundExpression node,
+            TypeSymbol type1,
+            Conversion conv1,
+            TypeSymbol type2,
+            Conversion conv2,
+            ref HashSet<DiagnosticInfo> useSiteDiagnostics,
+            out bool okToDowngradeToNeither,
+            int betterConversionTargetRecursionLimit = 100)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: This is the other location the depth is used, it should be a constant.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 1 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: This current pattern makes it easy to forget to pass the recursion thus rendering it useless.  If the separation between outside and inside overload resolution is teh goal then divide the methods:

1. BetterConversionTarget: called outside overload resolution, no need for a recursion argument.
2. BetterConversionTargetCore: called inside, pass the recursion limit.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: Will do.

**Comment by gafter** [comment]
@jaredpar Your comments have been addressed in a new delta on this PR.

@khyperia @dotnet/roslyn-compiler Can I get a second review, please?

**Comment by jaredpar** [comment]
👍

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs [hunk 3]
```diff
@@ -8401,5 +8401,102 @@ class D
...
+";
+            CreateCompilationWithMscorlib(source).VerifyDiagnostics(
+                // (11,9): error CS0121: The call is ambiguous between the following methods or properties: 'Program.Method(Del1)' and 'Program.Method(Del2)'
+                //         Method(() => null);
+                Diagnostic(ErrorCode.ERR_AmbigCall, "Method").WithArguments("Program.Method(Del1)", "Program.Method(Del2)").WithLocation(11, 9)
+                );
+        }
+
+        [Fact, WorkItem(12061, "https://github.com/dotnet/roslyn/issues/12061")]
+        public void RecursiveBetterBetterness02()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs]: Is this test different from the one above?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs]: Yes: in this one, the delegate types return each other.

**Comment by gafter** [comment]
@jaredpar Please let me know if you are happy with the PR in this form.

**Comment by jaredpar** [comment]
@gafter did something change since my :+1: yesterday?

**Comment by gafter** [comment]
Oh, for some reason didn't notice it. Thanks!

**Pull request merged**